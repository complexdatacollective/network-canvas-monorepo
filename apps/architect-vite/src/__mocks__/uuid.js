/* eslint-env jest */

const uuid = jest.fn(() => '809895df-bbd7-4c76-ac58-e6ada2625f9b');

export default uuid;
