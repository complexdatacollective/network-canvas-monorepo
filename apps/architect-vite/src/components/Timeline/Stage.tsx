import cx from "classnames";
import { DeleteIcon } from "lucide-react";
import { motion } from "motion/react";
import { useRef } from "react";
import { Button } from "~/lib/legacy-ui/components";
import getAbsoluteBoundingRect from "~/utils/getAbsoluteBoundingRect";
import EditStageButton from "./EditStageButton";

const findPos = (node) => {
	let curtop = 0;
	let curtopscroll = 0;
	let n = node;
	do {
		curtop += n.offsetTop;
		curtopscroll += n.offsetParent ? n.offsetParent.scrollTop : 0;
		n = n.offsetParent;
	} while (n.offsetParent);
	return curtop - curtopscroll;
};

const variants = {
	show: {
		scale: 1,
		opacity: 1,
		transition: { type: "spring", delay: 0.75 },
	},
	hide: {
		scale: 0,
		opacity: 0,
	},
};

type StageProps = {
	id: string;
	stageNumber: number;
	className?: string;
	label?: string;
	type: string;
	hasSkipLogic?: boolean;
	hasFilter?: boolean;
	onEditStage: (id: string, rect: { top: number; left: number; width: number; height: number }) => void;
	onDeleteStage: (id: string) => void;
};

const Stage = ({
	id,
	stageNumber,
	className = "",
	onEditStage,
	onDeleteStage,
	type,
	label = "",
	hasSkipLogic = false,
	hasFilter = false,
}: StageProps) => {
	const componentClasses = cx("timeline-stage", className);

	const previewRef = useRef();

	const handleEditStage = () => {
		if (!previewRef.current) {
			return;
		}

		const rect = getAbsoluteBoundingRect(previewRef.current);
		const find = findPos(previewRef.current);

		onEditStage(id, {
			...rect,
			top: find,
		});
	};

	return (
		<motion.div className={componentClasses} variants={variants}>
			<div className="timeline-stage__notch" onClick={handleEditStage}>
				{stageNumber}
			</div>
			<EditStageButton
				ref={previewRef}
				onEditStage={handleEditStage}
				type={type}
				label={label}
				hasSkipLogic={hasSkipLogic}
				hasFilter={hasFilter}
			/>
			<div className="timeline-stage__controls">
				<Button color="neon-coral" icon={<DeleteIcon />} onClick={() => onDeleteStage(id)} title="Delete stage">
					Delete stage
				</Button>
			</div>
		</motion.div>
	);
};

export { Stage as UnconnectedStage };

export default Stage;
