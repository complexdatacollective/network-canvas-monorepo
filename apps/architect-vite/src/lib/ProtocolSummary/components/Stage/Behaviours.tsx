import { map } from "lodash";
import MiniTable from "../MiniTable";
import { renderValue } from "../helpers";

const behaviorLabel = (behaviourValue: any, behaviourKey: string) => {
	switch (behaviourKey) {
		case "allowRepositioning":
			return { label: "Repositioning enabled", value: behaviourValue };
		case "automaticLayout":
			return { label: "Automatic layout enabled", value: behaviourValue.enabled };
		case "minNodes":
			return { label: "Minimum nodes on stage", value: behaviourValue };
		case "maxNodes":
			return { label: "Maximum nodes on stage", value: behaviourValue };
		case "freeDraw":
			return { label: "Freedraw enabled", value: behaviourValue };
		default:
			return { label: behaviourKey, value: behaviourValue };
	}
};

const behaviourRows = (behaviours: Record<string, any>) =>
	map(behaviours, (behaviourValue, behaviourKey) => {
		const labelValue = behaviorLabel(behaviourValue, behaviourKey);
		return [labelValue.label, renderValue(labelValue.value)];
	});

type BehavioursProps = {
	behaviours?: {
		allowRepositioning?: boolean;
		freeDraw?: boolean;
		[key: string]: any;
	} | null;
};

const Behaviours = ({ behaviours = null }: BehavioursProps) => {
	if (!behaviours) {
		return null;
	}

	return (
		<div className="protocol-summary-stage__behaviours">
			<div className="protocol-summary-stage__behaviours-content">
				<h2 className="section-heading">Behaviours</h2>
				<MiniTable rotated rows={behaviourRows(behaviours)} />
			</div>
		</div>
	);
};

export default Behaviours;
