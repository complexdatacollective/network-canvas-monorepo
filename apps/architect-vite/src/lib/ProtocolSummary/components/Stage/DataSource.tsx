import AssetBadge from "../AssetBadge";

type DataSourceProps = {
	dataSource?: string | null;
};

const DataSource = ({ dataSource = null }: DataSourceProps) => {
	if (!dataSource) {
		return null;
	}

	return (
		<div className="protocol-summary-stage__data-source">
			<div className="protocol-summary-stage__data-source-content">
				<h2 className="section-heading">DataSource</h2>
				<AssetBadge id={dataSource} link />
			</div>
		</div>
	);
};


export default DataSource;
