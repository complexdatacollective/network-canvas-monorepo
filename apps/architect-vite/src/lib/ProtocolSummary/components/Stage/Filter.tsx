import Rules from "../Rules";

type FilterProps = {
	filter: Record<string, unknown>;
};

const Filter = ({ filter }: FilterProps) => {
	if (!filter) {
		return null;
	}

	return (
		<div className="protocol-summary-stage__filter">
			<Rules filter={filter} />
		</div>
	);
};

export default Filter;
