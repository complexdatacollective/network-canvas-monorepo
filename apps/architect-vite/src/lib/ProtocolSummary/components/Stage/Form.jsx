import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import Markdown from '@codaco/ui/lib/components/Fields/Markdown';
import SummaryContext from '../SummaryContext';
import MiniTable from '../MiniTable';
import Variable from '../Variable';
import { getVariableMeta } from '../helpers';

const Form = ({ form }) => {
  const {
    index,
  } = useContext(SummaryContext);

  if (!form) { return null; }

  const fieldRows = form.fields && form.fields.map(({ prompt, variable }) => {
    const meta = getVariableMeta(index, variable);

    return ([
      <Variable id={variable} />,
      meta.component,
      <Markdown label={prompt} />,
    ]);
  });

  return (
    <div className="protocol-summary-stage__form">
      <div className="protocol-summary-stage__form-content">
        <h2 className="section-heading">Form</h2>
        { form.title && (
          <h4>
            Title:
            {' '}
            {form.title}
          </h4>
        )}
        <MiniTable
          wide
          rows={[
            ['Variable', 'Component', 'Prompt'],
            ...fieldRows,
          ]}
        />
      </div>
    </div>
  );
};

const FieldType = PropTypes.shape({
  prompt: PropTypes.string.isRequired,
  variable: PropTypes.string.isRequired,
});

Form.propTypes = {
  form: PropTypes.shape({
    title: PropTypes.string,
    fields: PropTypes.arrayOf(FieldType).isRequired,
  }),
};

Form.defaultProps = {
  form: null,
};

export default Form;
