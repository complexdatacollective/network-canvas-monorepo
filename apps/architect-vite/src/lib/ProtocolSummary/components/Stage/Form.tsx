import { useContext } from "react";
import Markdown from "~/components/Form/Fields/Markdown";
import MiniTable from "../MiniTable";
import SummaryContext from "../SummaryContext";
import Variable from "../Variable";
import { getVariableMeta } from "../helpers";

type FormFieldType = {
	prompt: string;
	variable: string;
};

type FormProps = {
	form?: {
		title?: string;
		fields?: FormFieldType[];
	} | null;
};

const Form = ({ form = null }: FormProps) => {
	const { index } = useContext(SummaryContext);

	if (!form) {
		return null;
	}

	const fieldRows = form.fields?.map(({ prompt, variable }) => {
		const meta = getVariableMeta(index, variable);

		return [<Variable id={variable} />, meta.component, <Markdown label={prompt} />];
	});

	return (
		<div className="protocol-summary-stage__form">
			<div className="protocol-summary-stage__form-content">
				<h2 className="section-heading">Form</h2>
				{form.title && <h4>Title: {form.title}</h4>}
				<MiniTable wide rows={[["Variable", "Component", "Prompt"], ...fieldRows]} />
			</div>
		</div>
	);
};

export default Form;
