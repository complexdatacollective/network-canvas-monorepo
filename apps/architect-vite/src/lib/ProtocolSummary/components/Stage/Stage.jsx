import { get, isEmpty, sortBy } from "es-toolkit/compat";
import PropTypes from "prop-types";
import React, { useContext } from "react";
import interfaceImage from "~/images/timeline";
import DualLink from "../DualLink";
import EntityBadge from "../EntityBadge";
import MiniTable from "../MiniTable";
import SummaryContext from "../SummaryContext";
import Behaviours from "./Behaviours";
import DataSource from "./DataSource";
import Filter from "./Filter";
import Form from "./Form";
import InterviewScript from "./InterviewScript";
import IntroductionPanel from "./IntroductionPanel";
import Items from "./Items";
import PageHeading from "./PageHeading";
import Panels from "./Panels";
import Presets from "./Presets";
import Prompts from "./Prompts";
import QuickAdd from "./QuickAdd";
import SkipLogic from "./SkipLogic";

const getInterfaceImage = (type) => get(interfaceImage, type);

const variablesOnStage = (index) => (stageId) =>
	index.reduce((memo, variable) => {
		if (!variable.stages.includes(stageId)) {
			return memo;
		}
		return [...memo, [variable.id, variable.name]];
	}, []);

const Stage = ({ configuration, id, label, stageNumber, type }) => {
	const { index } = useContext(SummaryContext);

	const stageVariables = sortBy(variablesOnStage(index)(id), [(variable) => variable[1].toLowerCase()]);

	return (
		<div className="protocol-summary-stage page-break-marker" id={`stage-${id}`}>
			<div className="protocol-summary-stage__heading">
				<div className="protocol-summary-stage__wrapper">
					<div className="protocol-summary-stage__summary">
						<div className="stage-label" data-number={stageNumber}>
							<h1>{label}</h1>
						</div>
						{configuration.subject && !isEmpty(stageVariables) && (
							<table className="protocol-summary-mini-table protocol-summary-mini-table--rotated">
								<tbody>
									{configuration.subject && (
										<tr>
											<td>Subject</td>
											<td>
												<EntityBadge
													small
													type={configuration.subject.type}
													entity={configuration.subject.entity}
													link
												/>
											</td>
										</tr>
									)}
									{!isEmpty(stageVariables) && (
										<tr>
											<td>Variables</td>
											<td>
												{stageVariables.map(([variableId, variable], i) => (
													<React.Fragment key={variableId}>
														<DualLink to={`#variable-${variableId}`}>{variable}</DualLink>
														{i !== stageVariables.length - 1 && ", "}
													</React.Fragment>
												))}
											</td>
										</tr>
									)}
								</tbody>
							</table>
						)}
					</div>
					<div className="protocol-summary-stage__preview">
						<div className="stage-image">
							<img src={getInterfaceImage(type)} alt="" />
						</div>
						{/* <h4>
              {type}
            </h4> */}
					</div>
				</div>
				{configuration.filter && (
					<div className="protocol-summary-stage__heading-section">
						<div className="protocol-summary-stage__heading-section-content">
							<h2 className="section-heading">Network Filtering</h2>
							<MiniTable rotated wide rows={[["Rules", <Filter filter={configuration.filter} />]]} />
						</div>
					</div>
				)}
				{configuration.skipLogic && (
					<div className="protocol-summary-stage__heading-section">
						<div className="protocol-summary-stage__heading-section-content">
							<h2 className="section-heading">Skip Logic</h2>
							<SkipLogic skipLogic={configuration.skipLogic} />
						</div>
					</div>
				)}
				<div className="protocol-summary-stage__content">
					<IntroductionPanel introductionPanel={configuration.introductionPanel} />
					<DataSource dataSource={configuration.dataSource} />
					<QuickAdd quickAdd={configuration.quickAdd} />
					<Panels panels={configuration.panels} />
					<Prompts prompts={configuration.prompts} />
					<Form form={configuration.form} />
					<Behaviours behaviours={configuration.behaviours} />
					<Presets presets={configuration.presets} />
					<PageHeading heading={configuration.title} />
					<Items items={configuration.items} />
					<InterviewScript interviewScript={configuration.interviewScript} />
				</div>
			</div>
		</div>
	);
};

Stage.propTypes = {
	id: PropTypes.string.isRequired,
	// eslint-disable-next-line react/forbid-prop-types
	configuration: PropTypes.object.isRequired,
	label: PropTypes.string.isRequired,
	stageNumber: PropTypes.number.isRequired,
	type: PropTypes.string.isRequired,
};

export default Stage;
