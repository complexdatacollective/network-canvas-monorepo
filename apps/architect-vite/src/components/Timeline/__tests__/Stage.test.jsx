import { shallow } from "enzyme";
import { v4 as uuid } from "uuid";
import { UnconnectedStage } from "../Stage";

jest.mock("@codaco/ui/lib/utils/CSSVariables");

const mockProps = {
	id: uuid(),
	stageNumber: 1,
	type: "Foo",
	onEditStage: () => {},
	onDeleteStage: () => {},
	onEditSkipLogic: () => {},
	onInsertStage: () => {},
};

describe("<Stage />", () => {
	it("can render", () => {
		const component = shallow(<UnconnectedStage {...mockProps} />);

		expect(component).toMatchSnapshot();
	});
});
