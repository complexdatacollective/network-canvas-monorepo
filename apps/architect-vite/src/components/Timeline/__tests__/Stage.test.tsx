import { render } from "@testing-library/react";
import { v4 as uuid } from "uuid";
import { describe, expect, it, vi } from "vitest";
import { UnconnectedStage } from "../Stage";

vi.mock("~/lib/legacy-ui/utils/CSSVariables");

const mockProps = {
	id: uuid(),
	stageNumber: 1,
	type: "Foo",
	onEditStage: () => {},
	onDeleteStage: () => {},
	onEditSkipLogic: () => {},
	onInsertStage: () => {},
};

describe("<Stage />", () => {
	it("can render", () => {
		const { container } = render(<UnconnectedStage {...mockProps} />);

		expect(container.firstChild).toBeInTheDocument();
	});
});
