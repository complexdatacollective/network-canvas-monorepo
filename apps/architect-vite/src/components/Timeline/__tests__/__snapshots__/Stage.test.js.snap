// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Stage /> can render 1`] = `
<ForwardRef(MotionComponent)
  className="timeline-stage"
  variants={
    Object {
      "hide": Object {
        "opacity": 0,
        "scale": 0,
      },
      "show": Object {
        "opacity": 1,
        "scale": 1,
        "transition": Object {
          "delay": 0.75,
          "type": "spring",
        },
      },
    }
  }
>
  <div
    className="timeline-stage__notch"
    onClick={[Function]}
  >
    1
  </div>
  <ForwardRef
    hasFilter={false}
    hasSkipLogic={false}
    label=""
    onEditStage={[Function]}
    type="Foo"
  />
  <div
    className="timeline-stage__controls"
  >
    <Button
      color="neon-coral"
      content=""
      disabled={false}
      icon={<Memo />}
      iconPosition="left"
      onClick={[Function]}
      size="small"
      title="Delete stage"
      type="button"
    >
      Delete stage
    </Button>
  </div>
</ForwardRef(MotionComponent)>
`;
